src/cache.rs

```diff
@@ -0,0 +1,283 @@
+        let schema = conn.current_schema();
+
+        let kw = NamespacedKeyword::new("foo", "bar");
+
+        conn.attribute_cache().add_to_cache(&mut sqlite, &schema,kw.clone()).expect("No errors on add to cache");
+        assert!(conn.attribute_cache().contains_key(&kw));
+        conn.attribute_cache().add_to_cache(&mut sqlite, &schema,kw.clone()).expect("No errors on add to cache");
+        assert!(conn.attribute_cache().contains_key(&kw));
+    }
+
+    #[test]
+    fn test_remove_from_cache() {
+        let mut sqlite = db::new_connection("").unwrap();
+        let mut conn = Conn::connect(&mut sqlite).unwrap();
+        let report = conn.transact(&mut sqlite, r#"[
+            {  :db/ident       :foo/bar
+               :db/valueType   :db.type/long },
+            {  :db/ident       :foo/baz
+               :db/valueType   :db.type/boolean }]"#).expect("transaction expected to succeed");
+        let report = conn.transact(&mut sqlite, r#"[
+            {  :foo/bar        100
+               :foo/baz        false },
+            {  :foo/bar        200
+               :foo/baz        true }]"#).expect("transaction expected to succeed");
+        let schema = conn.current_schema();
+
+        let kwr = NamespacedKeyword::new("foo", "bar");
+        let kwz = NamespacedKeyword::new("foo", "baz");
+
+        conn.attribute_cache().add_to_cache(&mut sqlite, &schema,kwr.clone()).expect("No errors on add to cache");
+        assert!(conn.attribute_cache().contains_key(&kwr));
+        conn.attribute_cache().add_to_cache(&mut sqlite, &schema,kwz.clone()).expect("No errors on add to cache");
+        assert!(conn.attribute_cache().contains_key(&kwz));
+
+        // test that we can remove an item from cache
+        conn.attribute_cache().remove_from_cache(&kwz).expect("No errors on remove from cache");
+        assert!(!conn.attribute_cache().contains_key(&kwz));
+    }
+
+    #[test]
+    fn test_remove_attribute_not_in_cache() {
+        let mut sqlite = db::new_connection("").unwrap();
+        let mut conn = Conn::connect(&mut sqlite).unwrap();
+        let report = conn.transact(&mut sqlite, r#"[
+            {  :db/ident       :foo/bar
+               :db/valueType   :db.type/long },
+            {  :db/ident       :foo/baz
+               :db/valueType   :db.type/boolean }]"#).expect("transaction expected to succeed");
+        let report = conn.transact(&mut sqlite, r#"[
+            {  :foo/bar        100
+               :foo/baz        false },
+            {  :foo/bar        200
+               :foo/baz        true }]"#).expect("transaction expected to succeed");
+        let schema = conn.current_schema();
+
+        let kw = NamespacedKeyword::new("foo", "baz");
+        let res = conn.attribute_cache().remove_from_cache(&kw);
+        match res.unwrap_err() {
+            Error(ErrorKind::CacheMiss(msg), _) => assert_eq!(msg, format!("{:?}", kw)),
```

Comment:
This is not what I think a `CacheMiss` means. I think you're raising this error when an attribute isn't being cached. I think of a cache miss as when an attribute _is_ being cached, but a particular entity's values are not present in the cache due to eviction.